### ISSUE TITLE
QWEN3-VL-8B 多图推理时 VIT 输出 NaN 导致生成无效 Token

### ISSUE THREAD

**Issue by PengRui1995** [NVIDIA/TensorRT-Edge-LLM#59]
Version：0.4.0
问题描述：在 QWEN3-VL-8B 端侧推理链路中，多图请求在图片数量和总视觉 token 较大时会出现生成异常。

**典型现象：**
- 单图正常 max_token_per_image=1024时
- `4 * 1080p` image 正常  
- `5 * 1080p 开始异常
- LLM 首 token 采样阶段出现无效 token：
Generated invalid token 151935 (>= tokenizer vocab 151669)

**更早的异常出现在 VIT 输出：**
VIT output embeddings: nan_count=5000
deepstack_features.2: nan=4096
Logits Top-5: logit=nan

这不是单纯的采样问题，而是视觉编码结果已经损坏，最终传导到 LLM。

---

### 问题链路

1. 图像预处理正常完成
2. VIT 推理完成，但输出 embedding 和部分 deepstack feature 出现 NaN
3. LLM prefill 阶段收到 NaN 的图像 embedding
4. logits 全部变成 NaN
5. 采样拿到超出 tokenizer vocab 的 token id

### 初步根因分析（通过gpt的，仅供参考，不对请指出）

**直接根因：**
QWEN3-VL 视觉分支在"多图拼接成一个超长 dense sequence"时，TensorRT 视觉 engine 在后半段 attention / block 执行中产生 NaN。

**为什么 4 图没问题，5 图开始有问题：**
- 4 图时总视觉长度 `4 * 1024 * 4 = 16384` (VIT seq_len)，未跨过风险边界
- 5 图时总视觉长度 `5 * 1024 * 4 = 20480`，超过边界
- 触发了 TensorRT dense attention 在大序列下的不稳定路径

**技术细节：**
- QWEN3-VL 中 `num_heads = 16`
- `seq_len = 16384` 时，attention 元素数 `16 * 16384 * 16384 = 2^32`
- 超过此值后，TensorRT/CUDA 的 dense attention tactic 出现索引溢出或偏移溢出

**多图 attention 实现：**
当前实现不是 ragged attention，而是：
- 先把多张图拼成一个大序列
- 再构造块对角 `attention_mask`
- 本质上是"单图 VIT 结果的拼接"，工程上一次大 launch 跑完

---

###初步解决方案（我们尝试过这样修改，似乎能解决问题，不对的请指正）

将 QWEN3-VL 的多图 VIT 推理从"多图拼成一个超长序列一次性送入 VIT"改为"逐图执行 VIT，再按原 token 顺序拼接 output / deepstack features"。

**实现要点：**
1. 预处理阶段记录每张图的元信息（原始尺寸、resize后尺寸、gridTHW、cuSeqlensStart/End、seqLength、tokenLength）
2. 新增 `inferQwen3PerImage()`，对 QWEN3-VL 多图请求遍历每张图单独执行 VIT
3. 保持单图路径和非 QWEN3-VL 模型路径不变

**为什么逐图执行在语义上等价：**
因为当前视觉实现里，多图 attention 本来就是块对角 mask，图 A 的 token 只看图 A，不存在真正的跨图视觉交互。

---

### 验证结果

- **单图验证**：VIT output embeddings 全有限，deepstack_features.0/1/2 全有限，LLM 正常生成
- **10 图验证**：`total_seq_len=34232`，远超之前触发异常的区间，由于逐图执行每次 VIT launch 的单图长度分别为 2852、3844、4032，全部处于稳定区间内

---

**Comment by nvluxiaoz** [comment]
Have you tried the latest 0.5.0 with ViT AttentionPlugin? 0.5.0 has better support for this scenario using ragged attention.

**Comment by nvluxiaoz** [comment]
This issue has been stale for > 3 months. Closing. If you run into the same issue with latest branch, please let us know. Thanks.
